File: src/runtime/boxes/BbAtp.cs
// BbAtp.cs — ATP: @var — write cursor position as integer into varname
// Mirrors src/runtime/boxes/bb_atp.c
//
// α: write Δ (cursor, 0-based) to varname; γ zero-width (always succeeds)
// β: ω (no retry — cursor capture is a side-effect, not a match)

namespace Snobol4.Runtime.Boxes;

public sealed class BbAtp : IByrdBox
{
    private readonly string _varname;

    // Injected by executor — writes integer-as-string to IdentifierTable
    public Action<string, string>? SetVar { get; set; }

    public BbAtp(string varname) { _varname = varname ?? ""; }

    public Spec Alpha(MatchState ms)
    {
        if (!string.IsNullOrEmpty(_varname))
            SetVar?.Invoke(_varname, ms.Cursor.ToString());
        return Spec.ZeroWidth(ms.Cursor);             // ATP_γ
    }

    public Spec Beta(MatchState ms) => Spec.Fail;    // ATP_ω
}

File: src/runtime/boxes/BbAtp.java
/**
 * BbAtp.java — ATP: @var — write cursor position Δ as integer into varname
 * Port of bb_atp.c / bb_atp.s
 *
 *   ATP_α:  if (varname) NV_SET(varname, Δ);
 *           ATP=spec(Σ+Δ,0);                    goto ATP_γ;
 *   ATP_β:                                      goto ATP_ω;
 */
class BbAtp extends BbBox {

    public interface IntSetter { void set(String varname, int value); }

    private final String    varname;
    private final IntSetter setter;

    public BbAtp(MatchState ms, String varname, IntSetter setter) {
        super(ms);
        this.varname = varname;
        this.setter  = setter;
    }

    @Override public Spec alpha() {
        if (varname != null && !varname.isEmpty())
            setter.set(varname, ms.delta);
        return new Spec(ms.delta, 0);                                          // ATP_γ
    }

    @Override public Spec beta() { return null; }                              // ATP_ω
}
